crates/assets/src/asset.rs

```diff
@@ -0,0 +1,901 @@
+		let reference_length = reference.len();
+		if reference_length < MINIMUM_REFERENCE_LENGTH {
+			Err(ReferenceError::TooShort)
+		} else if reference_length > MAXIMUM_REFERENCE_LENGTH {
+			Err(ReferenceError::TooLong)
+		} else {
+			Ok(())
+		}
+	}
+
+	const fn check_identifier_length_bounds(identifier: &[u8]) -> Result<(), IdentifierError> {
+		let identifier_length = identifier.len();
+		if identifier_length < MINIMUM_IDENTIFIER_LENGTH {
+			Err(IdentifierError::TooShort)
+		} else if identifier_length > MAXIMUM_IDENTIFIER_LENGTH {
+			Err(IdentifierError::TooLong)
+		} else {
+			Ok(())
+		}
+	}
+
+	/// Split the given input into its components, i.e., namespace, reference,
+	/// and identifier, if the proper separators are found.
+	#[allow(clippy::type_complexity)]
+	fn split_components(input: &[u8]) -> (Option<&[u8]>, Option<&[u8]>, Option<&[u8]>) {
+		let mut split = input.splitn(2, |c| *c == NAMESPACE_REFERENCE_SEPARATOR);
+		let (namespace, reference) = (split.next(), split.next());
+
+		// Split the remaining reference to extract the identifier, if present
+		let (reference, identifier) = if let Some(r) = reference {
+			let mut split = r.splitn(2, |c| *c == REFERENCE_IDENTIFIER_SEPARATOR);
+			// Split the reference further, if present
+			(split.next(), split.next())
+		} else {
+			// Return the old reference, which is None if we are at this point
+			(reference, None)
+		};
+
+		(namespace, reference, identifier)
+	}
+
+	/// A Slip44 asset reference.
+	/// It is a modification of the [CAIP-20 spec](https://github.com/ChainAgnostic/CAIPs/blob/master/CAIPs/caip-20.md)
+	/// according to the rules defined in the Asset DID method specification.
+	#[non_exhaustive]
+	#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, RuntimeDebug, Encode, Decode, MaxEncodedLen, TypeInfo)]
+	pub struct Slip44Reference(pub U256);
+
+	impl Slip44Reference {
+		/// Parse a UTF8-encoded decimal Slip44 asset reference, failing if the
+		/// input string is not valid.
+		pub(crate) fn from_utf8_encoded<I>(input: I) -> Result<Self, AssetIdError>
+		where
+			I: AsRef<[u8]> + Into<Vec<u8>>,
+		{
+			let input = input.as_ref();
+			check_reference_length_bounds(input)?;
+
+			let decoded = str::from_utf8(input).map_err(|_| ReferenceError::InvalidFormat)?;
+			let parsed = U256::from_dec_str(decoded).map_err(|_| ReferenceError::InvalidFormat)?;
+			// Unchecked since we already checked for maximum length and hence maximum value
+			Ok(Self(parsed))
+		}
+	}
+
+	impl TryFrom<U256> for Slip44Reference {
+		type Error = AssetIdError;
+
+		fn try_from(value: U256) -> Result<Self, Self::Error> {
+			// Max value for 64-digit decimal values (used for Slip44 references so far).
+			// TODO: This could be enforced at compilation time once constraints on generics
+			// will be available.
+			if value
+				<= U256::from_str_radix("9999999999999999999999999999999999999999999999999999999999999999", 10)
```

Comment:
Do you have an easy-to-understand proposal? I think starting playing around with arrays means that endianness and alignment must be taken into account, and I am afraid it would make the code a bit less readable. But I also agree that what is written right now is not the most efficient.

Reply:
You are probably right. 😞